File: backend/app/schemas/diagram.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


DiagramState = Literal[
    "pending",
    "spec_generating",
    "rendering",
    "rate_limited",
    "completed",
    "failed",
    "skipped",
    "stale",
]
DiagramMode = Literal["structured", "illustrative"]


class DiagramAssetResponse(BaseModel):
    id: UUID
    video_id: UUID
    course_id: UUID
    marker_index: int
    original_caption: str
    detailed_prompt: str | None
    alt_text: str | None
    render_mode: DiagramMode | None
    mermaid_source: str | None
    provider: str | None
    model: str | None
    state: DiagramState
    retry_at: datetime | None
    image_url: str | None = None
    width: int | None
    height: int | None
    revision: int
    error_message: str | None

    model_config = ConfigDict(from_attributes=True)


class DiagramStatusResponse(BaseModel):
    course_id: UUID
    discovered: int
    pending: int
    processing: int
    waiting: int
    completed: int
    failed: int
    skipped: int
    stale: int


class DiagramGenerateResponse(BaseModel):
    course_id: UUID
    discovered: int
    queued: int


class DiagramRegenerateRequest(BaseModel):
    prompt: str | None = Field(default=None, max_length=5000)
    mode: DiagramMode | None = None

File: frontend/src/types/diagram.ts
export type DiagramState =
  | "pending"
  | "spec_generating"
  | "rendering"
  | "rate_limited"
  | "completed"
  | "failed"
  | "skipped"
  | "stale";

export type DiagramMode = "structured" | "illustrative";

export interface DiagramAsset {
  id: string;
  video_id: string;
  course_id: string;
  marker_index: number;
  original_caption: string;
  detailed_prompt: string | null;
  alt_text: string | null;
  render_mode: DiagramMode | null;
  mermaid_source: string | null;
  provider: string | null;
  model: string | null;
  state: DiagramState;
  retry_at: string | null;
  image_url: string | null;
  width: number | null;
  height: number | null;
  revision: number;
  error_message: string | null;
}

export interface DiagramStatus {
  course_id: string;
  discovered: number;
  pending: number;
  processing: number;
  waiting: number;
  completed: number;
  failed: number;
  skipped: number;
  stale: number;
}

export interface DiagramGenerateResult {
  course_id: string;
  discovered: number;
  queued: number;
}
